### ISSUE TITLE
Memory leak when allocating arrays in a for loop

### ISSUE THREAD

**Issue by ctiennot** [google/jax#19429]
### Description

Hi there, 

The issue was surfaced in Numpyro initially: https://github.com/pyro-ppl/numpyro/issues/1699, it seems all memory is not properly released when allocating a new array in a for loop.

One reproducible example:

```python
import jax, jaxlib;
import jax.numpy as jnp
import os
import psutil
import gc

process = psutil.Process(os.getpid())

print(jax.__version__, jaxlib.__version__)

@jax.jit
def fn(x):
  return jnp.sum(jnp.sin(x))

for i in range(0, 1_001):
    res = fn(jnp.zeros(i))

    jax.clear_caches()

    del res
    for x in jax.live_arrays():
        x.delete()

    gc.collect()

    if i % 100 == 0:
      print(f"{i=}: {process.memory_info().rss / 1024 / 1024}")
```

Gives:
```
0.4.23 0.4.23
i=0: 153.28125
i=100: 179.421875
i=200: 182.40625
i=300: 183.84375
i=400: 186.25
i=500: 187.71875
i=600: 189.703125
i=700: 192.046875
i=800: 194.734375
i=900: 196.890625
i=1000: 198.265625
```

More examples are available here: https://github.com/pyro-ppl/numpyro/issues/1699#issuecomment-1877573525

Thanks!

### What jax/jaxlib version are you using?

0.4.23, 0.4.23

### Which accelerator(s) are you using?

CPU

### Additional system info?

1.26.0 3.10.12 | packaged by conda-forge | (main, Jun 23 2023, 22:41:52) [Clang 15.0.7 ] uname_result(system='Darwin', node='Clements-MacBook-Pro-2.local', release='23.2.0', version='Darwin Kernel Version 23.2.0: Wed Nov 15 21:53:18 PST 2023; root:xnu-10002.61.3~2/RELEASE_ARM64_T6000', machine='arm64')

**Comment by jacoblubecki** [comment]
The `Array` implementation uses a top-level lru-caching function here:
https://github.com/google/jax/blob/0b542ff585b30b24aceadd12a2335b0fa26b8209/jax/_src/array.py#L124

There may be other places with similar caches that I didn't go deep enough to dig up... but these would cause some amount of memory growth independent of the jax-specific caching (e.g. anything that would be cleared by `clear_caches`) and a garbage collection would not free these LRU caches either.

I also tried your test code without the JIT annotation and observed the same linear growth which makes me think that at least that is not the problem... So I experimented a bit.

If you replace:
```
res = fn(jnp.zeros(i))
```

... with:

```
res = fn(jnp.zeros(5))
```

You will still observe some memory growth, but the growth is more stable.

For example, I tested by increasing the iteration count to 5000 (larger than the cache limit) and ran it with a fixed array size and got:
```
0.4.24 0.4.24
i=0: 127.640625
i=500: 138.46875
i=1000: 140.53125
i=1500: 140.59375
i=2000: 145.265625
i=2500: 145.28125
i=3000: 147.28125
i=3500: 147.953125
i=4000: 147.953125
i=4500: 147.96875
i=5000: 148.0
```

This is sort of the behavior you would expect from an LRU cache, so I would only be concerned if the memory growth is unbounded.

You mentioned in the other ticket:

> my model has varying input sizes and needs to reestimate all parameters regularly

This would mean you are likely to have a lot of cache misses and would end up fully utilizing the cache.
